src/main/scala/dpla/ingestion3/reports/ThumbnailReport.scala

```diff
@@ -0,0 +1,212 @@
+    * Images are items where sourceResource.`type` = image
+    */
+  def missingReport(ds: Dataset[DplaMapData], spark: SparkSession): DataFrame = {
+    import spark.implicits._
+
+    val thumbnailData: Dataset[MissingThumbnail] = ds.map(dplaMapData => {
+
+      val hasImageType = dplaMapData.sourceResource.`type`
+        .filter {
+          _.toLowerCase() == "image"
+        }.nonEmpty
+
+      val hasPreview = previewUri(dplaMapData).nonEmpty
+
+      MissingThumbnail(
+        localUri(dplaMapData),
+        dplaUri(dplaMapData),
+        hasImageType,
+        hasPreview)
+    })
+
+    thumbnailData.select("localUri", "dplaUri")
+      .filter("hasImageType = TRUE")
+      .filter("hasPreview = FALSE")
+  }
+
+  /**
+    * Get thumbnail dimensions, and any errors that occur when attempting a
+    * request the thumbnail endpoint.
+    * Warning: This process takes a long time to run.
+    */
+  def dimensionsReport(ds: Dataset[DplaMapData], spark: SparkSession): DataFrame = {
+    import spark.implicits._
+
+    val thumbnailData: Dataset[ThumbnailDimensions] = ds.map(dplaMapData => {
+
+      val uri: Option[URI] = previewUri(dplaMapData)
+
+      val preview: Option[String] = uri match {
+        case Some(uri) => Some(uri.toString)
+        case None => None
+      }
+
+      val dimensions: Option[Dimensions] = uri match {
+        case Some(uri) => Some(getDimensions(uri))
+        case None => None
+      }
+
+      ThumbnailDimensions(
+        localUri(dplaMapData),
+        dplaUri(dplaMapData),
+        preview,
+        dimensions
+      )
+    })
+
+    thumbnailData.select("localUri", "dplaUri", "preview", "dimensions.height",
+      "dimensions.width", "dimensions.error")
+  }
+
+  /**
+    * @param dplaMapData
+    * @return String, the DPLA URI for an item
+    */
+  def localUri(dplaMapData: DplaMapData): String =
+    dplaMapData.edmWebResource.uri.toString
+
+  /**
+    * @param dplaMapData
+    * @return String, the provider URI for an item
+    */
+  def dplaUri(dplaMapData: DplaMapData): String =
+    dplaMapData.oreAggregation.uri.toString
+
+  /**
+    * @param dplaMapData
+    * @return Option[URI], the thumbnail URI
+    */
+  def previewUri(dplaMapData: DplaMapData): Option[URI] =
+    dplaMapData.oreAggregation.preview match {
+      case Some(preview) => Some(preview.uri)
+      case None => None
+    }
+
+  /**
+    * Parse the height and width from the thumbnail image.
+    *
+    * @param uri URI for the thumbnail
+    * @return Dimensions case class containing the height and width, or an error
+    *         message if an error occurs.
+    */
+  def getDimensions(uri: URI): Dimensions = {
+    bufferedImg(uri) match {
+      case Success(bimg) =>
+        // Checking for null here creates a codacy error, but since BufferedImage
+        // is a Java class, it can legitimately be null.
+        if (bimg == null) Dimensions(error = Some("Buffered image is null."))
```

Comment:
Stupid Java! I'm thinking of writing a function somewhere that just returns `null` so we can use it instead of `null` everywhere else and have only one warning about it.

Reply:
Clever idea!